Add kamino.auto.crontab feature

Michael-Sinz opened the pull request:
Kamino is designed to be lights-out operation.  As such, it can
be installed as a crontab that runs the auto-update process on
a regular basis.

In most of our clusters, we run this automatic update process on
a daily basis.  Since OS updates usually don't happen daily, this
ends up just checking and noting that there is nothing to do.

When an OS update does happen (or the age limit on the prior image
is reached) then the auto-update process does what it needs to do.

This allows the whole thing to be running in clusters without
intervention.

The schedule could easily be run more often as the cronjob is set
up to not run while a prior instance is running.  I run my test
clusters with this turned to running every few minutes just to
stress test the logic and to quickly notice a condition where a
new image is needed (just to reduce testing time).

This fixes #44

Enhanced smoketest3.sh a bit to make it do what I needed.

The new smoketest4.sh is just like smoketest3.sh only it uses the
cronjob (with an "every minute" schedule) to run the cronjob and
then uninstalls after it completes so as to not continually rerun.
It is just a smoketest of the chart...

Michael-Sinz committed:
Add kamino.auto.crontab feature

Kamino is designed to be lights-out operation.  As such, it can
be installed as a crontab that runs the auto-update process on
a regular basis.

In most of our clusters, we run this automatic update process on
a daily basis.  Since OS updates usually don't happen daily, this
ends up just checking and noting that there is nothing to do.

When an OS update does happen (or the age limit on the prior image
is reached) then the auto-update process does what it needs to do.

This allows the whole thing to be running in clusters without
intervention.

The schedule could easily be run more often as the cronjob is set
up to not run while a prior instance is running.  I run my test
clusters with this turned to running every few minutes just to
stress test the logic and to quickly notice a condition where a
new image is needed (just to reduce testing time).

This addresses issue #44

Enhanced smoketest3.sh a bit to make it do what I needed.

The new smoketest4.sh is just like smoketest3.sh only it uses the
cronjob (with an "every minute" schedule) to run the cronjob and
then uninstalls after it completes so as to not continually rerun.
It is just a smoketest of the chart...

Michael-Sinz committed:
Added some extra help for the cronjob schedule value

Michael-Sinz commented on helm/vmss-prototype/values.yaml:
```diff
@@ -72,8 +72,24 @@ kamino:
+      #schedule: "42 * * * *"
+
+      # A reasonable default is to do it every day.  (Obviously, only if enalbed)
+      schedule: "42 0 * * *"
```
Yes, a bit of humor from the ultimate guide to life's travels :smirk:

Michael-Sinz commented:
The check is failing due to some azure login at the start (unrelated to this change, I believe)

Michael-Sinz committed:
Add some code to deal with the machine-id

This fixes issue #22 by zapping the /etc/machine-id such that
new nodes, on boot, will define their own unique value rather than
retain the value from the prior instance.

This required a few extra helm chart changes.

Since this is not strictly required, we just "try" to do this but
if it fails, we move on and create the disk image like we did before.

Michael-Sinz committed:
Merge branch 'machine-id' into main

Michael-Sinz commented:
Note that I also now merged in the fix for the fixed machine-id value (it now gets generated for each new instance rather than remembered from the source node instance)

This fix is, as far as I can tell, mostly optional.  While machine-id is defined as being unique, it is only "randomly unique" and has not had any noticeable impact on our systems when they were not unique.

This change make this issue go away at the cost of a bit of time to run a pod on the donor node before it is deallocated (but after it is cordoned)

amrmahdi commented on helm/vmss-prototype/README.md:
```diff
@@ -202,6 +202,18 @@ Kamino can follow some rules for automatic node selection.  These are the parame
+  - When set to true, will run auto-update job on a periodic basis.  This requires that you use the `kamino.targetVMSS` setting since cronjobs only make sense in automatic mode.
+  - This is how we recommend running Kamino
+
+- `kaminog.auto.cronjob.schedule` (optional - defaults to "`42 0 * * *`")
```
```suggestion
- `kamino.auto.cronjob.schedule` (optional - defaults to "`42 0 * * *`")

```

Michael-Sinz replied:
Thanks

amrmahdi commented on vmss-prototype/vmss-prototype:
```diff
@@ -352,6 +352,98 @@ def get_vmss_images(subscription, resource_group, sig_name, image_definition):
+    # This is a pod yaml for our image tweak we will do on the target node
+    # that will be our new prototype image.
+    yaml_text = ("apiVersion: v1\n"
+                 "kind: Pod\n"
```
why not a job?

amrmahdi commented on vmss-prototype/vmss-prototype:
```diff
@@ -352,6 +352,98 @@ def get_vmss_images(subscription, resource_group, sig_name, image_definition):
+                 "  - name: skyman-acr\n"
+                 "  nodeSelector:\n"
+                 "    kubernetes.io/hostname: '{3}'\n"
+                 "  priorityClassName: skyman-cluster-daemonset\n"
```
does this exist?

Michael-Sinz replied:
Oops, in my clusters :-)
Will remove that.

amrmahdi commented on vmss-prototype/vmss-prototype:
```diff
@@ -352,6 +352,98 @@ def get_vmss_images(subscription, resource_group, sig_name, image_definition):
+    # no condition for completed so the kubectl wait command can not work
+    # here and trying to use it for Ready ends up in a race with completion.
+    # Thus, we use the logs method.  Not as pretty but it works.
+    run(['kubectl', 'logs',
```
would be cleaner with a job.

Michael-Sinz replied:
Maybe it is worth it for this case to use a job and the overhead in kubernetes and let the job do all of this (similar) work when I ask to wait for the job to complete.  I am unsure it buys us much.

Michael-Sinz replied:
For now, I would rather leave it like this.  We can look back on this later but jobs do not add much over a pod when I am really just needing to make a one-shot pod for a specific node that is about to go away.

Michael-Sinz committed:
Forgot to remove an entry that is only for my clusters

Michael-Sinz committed:
Forgot another skyman cluster specific setting...

Michael-Sinz committed:
Clean up the logging a bit

Also add a longer timeout to the smoketest, just in case there
is some extended period of image pull

Michael-Sinz committed:
Update helm/vmss-prototype/README.md

Co-authored-by: Amr Hanafi <amramahdi@gmail.com>

Michael-Sinz committed:
minor formatting

Michael-Sinz committed:
Some enhanced wording describing the cronjob benefit

Michael-Sinz committed:
Make the mount even more restricted

This makes the tweaks mount restricted to the file /etc/machine-id
such that it only has access to that.  A much cleaner mount.
